File: solutions/typescript/063-binary-tree-level-order-traversal.ts
/**
 * Problem 63: Binary Tree Level Order Traversal (LeetCode 102)
 * Difficulty: Med
 * Language: TypeScript
 */
type TreeNode = { val: number; left: TreeNode | null; right: TreeNode | null };

function levelOrder(root: TreeNode | null): number[][] {
  if (!root) return [];
  const bfs = (queue: TreeNode[]): number[][] => {
    if (queue.length === 0) return [];
    const vals = queue.map(n => n.val);
    const next = queue.flatMap(n =>
      [n.left, n.right].filter((c): c is TreeNode => c !== null)
    );
    return [vals, ...bfs(next)];
  };
  return bfs([root]);
}

File: solutions/rust/063-binary-tree-level-order-traversal.rs
/**
 * Problem 63: Binary Tree Level Order Traversal (LeetCode 102)
 * Difficulty: Med
 * Language: Rust
 */
fn level_order(root: &Option<Box<TreeNode>>) -> Vec<Vec<i32>> {
    let mut result = vec![];
    let mut queue: Vec<&TreeNode> = match root {
        None => return result,
        Some(n) => vec![n.as_ref()],
    };
    while !queue.is_empty() {
        let vals: Vec<i32> = queue.iter().map(|n| n.val as i32).collect();
        result.push(vals);
        queue = queue.iter().flat_map(|n| {
            let mut ch = vec![];
            if let Some(ref l) = n.left { ch.push(l.as_ref()); }
            if let Some(ref r) = n.right { ch.push(r.as_ref()); }
            ch
        }).collect();
    }
    result
}
